### DISCUSSION TITLE
Changelog for CT updates

### DISCUSSION THREAD

**Discussion by Mark2Mark** [coldtype/coldtype#106] [Q&A]
Hi, is there a good way to find out about changes that potentially broke code after a CT update?

I updated CT now after almost a quarter year of not using it. Now many of my animations don’t work anymore due to changes in the API.

E.g. I found that the `Loop` class has been [eliminated](https://github.com/goodhertz/coldtype/commit/a95b9c308c58efbda21ace05f6955425ae39c416), but I cannot simply fix my code to work. I am digging around in the commit history to try to find the changes. But even if I manage to find some puzzle pieces or cluees and get the animation running, it does not look like intended (like the rendered clip from back then).
I had something that was kind of similar to the pre-commit-state of [this](https://github.com/goodhertz/coldtype/commit/265176bf39b67855fd23c2a3a2f5fb1be2b11ce3).

Another example of things that broke: I have used these kind of things from an asciiTimeline
```py
phase = f.a.t.current_phase(f.i)
state = phase.calc_state(states)
weight = phase.calc_state(weights)
```
But those are gone, too. I cannot even find `current_phase` in the entire repo now as a search term.

I tried to rewrite my code to the newer asciiTimeline API, but from the examples it’s a bit tricky to make sense. Those are so great examples, but they are also all so different and rather cryptic or hard to see what’s going on there. 
Is there some documentation about the asciiTimeline?

Just my 2 cents. Maybe there’s something to be useful to be discussed :)

So long, still all my biggest appreciation for your amazing work!

**Comment by stenson** [comment]
Hi Mark — thanks for posting, and apologies for the broken code. I have to admit this is part of the reason I have "alpha-quality software" disclaimer on the homepage, as I still think of the API as subject-to-change until a 1.0 release. (Also I didn't really know anyone was using any of those classes, and I refactored them in order to clean up a ton of duplicate functionality that had developed.)

I do think sticking with an older version of Coldtype is probably the best solution for an existing animation, or I'd be happy to take a look at your source code to see how difficult it'd be to transition to a newer API in Coldtype itself.

All that said, with a little bit of patching, most of the old helper classes (like `Loop`) should be able to be revived. Here's an example of using an old version of the Loop class with the latest version of Coldtype:

```python
import math
from coldtype.interpolation import loopidx, interp_dict
from coldtype.time.timeable import Timeable
from coldtype.time.timeline import Timeline


class LoopPhase():
    """
    Clip-like representation of a segment of a Loop, most likely retrieved from :meth:`~coldtype.time.loop.Loop.current_phase`
    * Standard ``Timeable`` available as ``.t``
    * ``.is_transition`` indicates if phase represents a transition
    """
    def __init__(self, is_transition:bool, i:int, timeable:Timeable):
        self.t = timeable
        self.i = i
        self.is_transition = is_transition
    
    def calc_state(self, states, e="eeio"):
        """
        Calculate progress of the loop phase, based on an easing fuction, ``e``
        * ``e=`` takes a mnemonic as enumerated in :func:`coldtype.time.easing.ease`
        """
        state = loopidx(states, self.t.idx)
        if self.is_transition:
            next_state = loopidx(states, self.t.idx+1)
            easeable = self.t.at(self.i)
            e = easeable.e(e, 0, to1=1)
            #e = self.t.progress(self.i, easefn=e, to1=1).e
            state = interp_dict(e, state.copy(), next_state)
        return state


class Loop(Timeline):
    """
    Construct for quickly developing animations based on loop
    
    Can be used as a ``timeline=`` for an ``@animation``
    """
    def __init__(self, duration, transition_length, states,
        loop=1, fps=30, storyboard=[0]):

        self.timeables = []
        self.transitions = []

        super().__init__(duration, fps=fps, storyboard=storyboard)

        try:
            segment_count = len(states)
            self.states = states
        except:
            segment_count = states
            self.states = None

        idx = 0
        segment_count = int(math.floor(segment_count))
        segment_length = duration / segment_count
        segment_frames = int(math.floor(segment_length))
        leftover = duration % segment_frames

        for x in range(0, self.duration, segment_frames):
            if idx >= segment_count:
                continue

            start = x
            end = x + segment_frames
            if idx == segment_count-1:
                end += leftover
            
            self.timeables.append(Timeable(start, end, index=idx))

            if idx < segment_count - 1:
                tstart = end-transition_length
                tend = end+transition_length
                self.transitions.append(Timeable(tstart, tend, index=idx))
            
            idx += 1
        
        if loop:
            self.transitions.insert(0, Timeable(-transition_length, transition_length, index=-1))
            self.transitions.append(Timeable(self.duration-transition_length, self.duration+transition_length, index=idx-1))
    
    def jumps(self):
        js = self._jumps
        for t in self.timeables:
            js.insert(-1, t.start+int(t.duration/2))
        return js
    
    def current_clip(self, i):
        for t in self.timeables:
            if t.now(i):
                return t

    def current_transition(self, i):
        for t in self.transitions:
            if t.now(i):
                return t

    def current(self, i):
        clip = self.current_clip(i)
        transition = self.current_transition(i)
        if transition:
            return True, transition
        else:
            return False, clip

    def current_on_loop(self, i):
        if i < 0:
            i = self.duration + i
        is_transition, timeable = self.current(i)
        return is_transition, i, timeable
    
    def current_phase(self, i) -> LoopPhase:
        """
        Get current phase of the Loop as a ``LoopPhase``
        """
        return LoopPhase(*self.current_on_loop(i))
    
    def current_state(self, i, e="eeio") -> LoopPhase:
        """
        Get current state of the Loop as a dict (provided states were passed to init)
        """
        return LoopPhase(*self.current_on_loop(i)).calc_state(self.states, e)
    
from coldtype import *

states = [
    dict(wdth=0, rotate=0, tu=300),
    dict(wdth=1, rotate=15, tu=-80),
    dict(wdth=0.5, rotate=-270, tu=330),
    dict(wdth=1, rotate=25, tu=50)
]

loop = Loop(70, 9, len(states))

@animation((1080, 1080/2), timeline=loop, bg=1, render_bg=1)
def render(f):
    state = f.a.t.current_phase(f.i).calc_state(states)
    return (StSt("COLDTYPE", Font.ColdtypeObviously(),
        150, fill=0, **state, r=1, leading=80)
        .align(f.a.r)
        .f(0))
```

**Comment by Mark2Mark** [comment]
Hi Rob, yeah, don’t worry about the broken code. I am aware of the Alpha state and the disclaimer, so that’s more than fair of you. I actually also enjoy seeing Coldtype’s development as well. It’s easier when one has the time to regularly check in, which I had not, unfortunately. But happy to be back.

I’d also be happy to help with documentation and things alike. If there’s somehow a way I can contribute, feel free to ping me.

For the time being, I think I will just collect every little puzzle piece that I find in your examples, and videos with a little note/explaination for myself. Maybe at some point that can be used publicly.

Also heads up for the refactoring towards even more clean and less dupe code 👏

I think my broken animations are not worth the work, I’ll rather look forward and work with the newer classes/functions.

Thanks again for this amazing piece! I am in total awe each time I dig in the source codebase.
